File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
public class Solution {

    public static int[] parent;

    public static int solution(int n, int[][] wires) {
        int answer = n;
        parent = new int[n + 1];

        for (int exclude = 0; exclude < wires.length; exclude++) {
            for (int i = 0; i <= n; i++) {
                parent[i] = i;
            }

            for (int j = 0; j < wires.length; j++) {
                if (exclude == j) {
                    continue;
                }

                int x = wires[j][0];
                int y = wires[j][1];
                union(x, y);
            }

            int count = 0;
            for (int i = 1; i < parent.length; i++) {
                if (find(parent[i]) == 1) {
                    ++count;
                } 
            }

            answer = Math.min(answer, Math.abs(n - 2 * count));
        }

        return answer;
    }

    public static void union(int x, int y) {
        x = find(x);
        y = find(y);

        if (x == y) {
            return;
        }

        if (x < y) {
            parent[y] = x;
        } else {
            parent[x] = y;
        }
    }

    public static int find(int x) {
        if (x == parent[x]) {
            return parent[x];
        }
        return parent[x] = find(parent[x]);
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>

using namespace std;

int CountNode(int curNode, vector<vector<bool>>& graph, vector<bool>& visited)
{
    int nodeCount = 1;
    for(int i = 1; i < graph.size(); ++i)
    {
        if (graph[curNode][i] == false)
            continue;
        
        if (visited[i])
            continue;
        
        visited[i] = true;
        nodeCount += CountNode(i, graph, visited);
    }
    
    return nodeCount;
}

int solution(int n, vector<vector<int>> wires) {
    int answer = n;
    
    vector<vector<bool>> graph (n + 1, vector<bool>(n + 1, false));
    
    for(int i = 0; i < wires.size(); ++i)
    {
        int n1 = wires[i][0];
        int n2 = wires[i][1];
        
        graph[n1][n2] = true;
        graph[n2][n1] = true;
    }
    
    for(int i = 0; i < wires.size(); ++i)
    {
        int n1 = wires[i][0];
        int n2 = wires[i][1];
        
        graph[n1][n2] = false;
        graph[n2][n1] = false;
        
        vector<bool> visited (n+1, false);
        visited[n1] = true;
        int nodeCount = CountNode(n1, graph, visited);
        
        answer = min(answer, abs(n - nodeCount - nodeCount));
        
        graph[n1][n2] = true;
        graph[n2][n1] = true;
    }
    
    return answer;
}